Update channel topic when Test Day starts and ends

galgeek opened the pull request:
hi @whimboo !

with apologies for the untidy commit list, suspecting this requires re-forking the repo to fix, and not sure what that would do to existing pull requests

whimboo commented on bot.js:
```diff
@@ -10,12 +10,15 @@ var ircServer = config.server,
       autoRejoin: config.autoRejoin,
     },
     client = new irc.Client(ircServer, nick, options),
+    botChannel = config.channels[0],
```
Just name it `channel`. It won't be anything else than the bot. :)

galgeek replied:
node irc uses `channel` throughout its documentation for function parameters, so it's a little confusing for anyone who's looking at that documentation to use `channel` as a separate variable name. I've made the change. I wonder what you'd prefer for the node irc function parameters, which I'm calling `channl` for now.

whimboo replied:
By default we use an `a` prefix for function parameters across all of our repositories. It would be good to refactor the code appropriate, but lets move this out to another issue. This could even be a good first mentored bug for someone. Lets skip that for now.

whimboo commented on bot.js:
```diff
@@ -53,15 +56,23 @@ function checkTestDay() {
 
+client.addListener('topic', function (channel, channelTopic, nick) {
+  if (!testDay && (channel === botChannel)){
+    topic_backup = channelTopic; // save a non-Test Day topic to restore after Test Day
```
nit: Please move the comment to the line before.

whimboo commented:
Also I think at the moment the bot waits for an activity in the channel before it updates the topic of the channel when the testday is about to start. Shouldn't this be handled with a timer event or such?

galgeek commented:
hi @whimboo 

a biggish update, now using setTimeout to start and end test day, and other updates per your recent feedback

thank you!

whimboo commented on bot.js:
```diff
@@ -51,19 +55,25 @@ function resetData() {
 
-client.addListener('join', function(channel, who){
-  checkTestDay();
+client.addListener('topic', function (channl, channlTopic, nick) {
```
Please revert this change to the parameter names. Stripping of chars should never be an option.

whimboo commented on bot.js:
```diff
@@ -51,19 +55,25 @@ function resetData() {
+    testDay = true;
+    resetData();
+    client.send('TOPIC', channel, topic);
+    timerID = setTimeout(checkTestDay, endTime - Date.now());
```
I'm not really a friend of setting timers for a long timespan. It should be ok for now, but in the future we might want to have a period like timer, which checks for all different kinds of tasks each minute or so.

whimboo replied:
Also keep in mind that this function is not called if the bot is joining IRC with an already set next testday. Currently we do not save the testday info, but that is something we have to do. Please file a new issue which can cover that.

whimboo commented on bot.js:
```diff
@@ -51,19 +55,25 @@ function resetData() {
 
 function checkTestDay() {
```
We should update the name of the method. It's not really a check anymore but more an `updateTestDayData`.

whimboo commented on bot.js:
```diff
@@ -168,8 +176,8 @@ client.addListener('pm', function(from, message){ // private messages to bot
       if (testDayAdmins.indexOf(from) >= 0){
-        testDay = false;
-        endTime = Date.now();
+        clearTimeout(timerID);
```
Please do that in the checkTestDay method.

whimboo commented on bot.js:
```diff
@@ -185,10 +193,19 @@ client.addListener('pm', function(from, message){ // private messages to bot
+          if (timerID !== 0){
+            clearTimeout(timerID);
+          }
+          timerID = setTimeout(checkTestDay, startTime - Date.now());
```
We cannot clear the timer here, especially if a testday is currently active.

galgeek replied:
The current code runs the :next command only when there's not a test day in progress.
Clearing the timer here allows us to reschedule a test day.

whimboo replied:
Sorry, but I do not see any check for an active testday in the :next callback. You only check if the given times are valid and then reset the timer. This will kill the timer to reset the topic when the currently active testday is over. What you want instead is to check in updateTestdayData() if a next testday is scheduled, and set the timer appropriately.

galgeek commented:
Hi @whimboo 

I've updated per your comments, including adding issue #43 save and reload test day schedule across bot restarts.

Note that current code accepts the :next admin command only when a test day is not in progress, so I've made no changes to `clearTimeout` there.

galgeek commented:
Hi @whimboo!

I've rebased and added test and qa config files.

I've left this version of the bot running using the test config, with a test Test Day scheduled for 6A-10A UTC. I'll check that it starts before I go to sleep, and maybe you'll have a chance to check that it's still running fine in the morning CET.

Thank you!

whimboo commented on config.json.test:
```diff
@@ -0,0 +1,10 @@
+    "autoRejoin": "true",
+    "admins": ["ashughes","whimboo","galgeek","bara"],
+    "helpers": ["ashughes","whimboo","bara"]
+}
```
Please remove that file from the patch. It's nice for testing but does not belong to the repository.

whimboo commented on config.json.qa:
```diff
@@ -0,0 +1,10 @@
+    "autoRejoin": "true",
+    "admins": ["ashughes", "whimboo"],
+    "helpers": ["ashughes"]
+}
```
Same for this file.

whimboo commented on bot.js:
```diff
@@ -195,8 +206,7 @@ client.addListener('pm', function(from, message) { // private messages to bot
         break;
       case ":stop":
         if (testDay) {
-          testDay = false;
```
I don't really like that this got moved into `updateTestDayData()` because it forces an early abort.

whimboo replied:
But that's necessary due to the current implementation of this method. See my above comment in what we should do later.

whimboo commented on bot.js:
```diff
@@ -52,21 +56,30 @@ function resetData() {
-      testDay = true;
-      resetData();
-    }
+    testDay = true;
```
This changes the status of the testday now each time the method get called. :/ It's ok for now but in the future this method should really only check if any action needs to be done depending on the current state of the testday.

galgeek commented:
Hi @whimboo!

I've removed the testing configs from the repo and updated .gitignore, and I've updated the PR.

I'm leaving any additional changes to updateTestDayData for later.

Thank you!

whimboo commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 .DS_Store
-
+config.json.test
+config.json.qa
```
Those are your own personal files on your machine. This is not for general use. So please remove.

galgeek commented:
Hi, @whimboo!

As we noted in IRC, I've updated this PR, backing out the .gitignore edits.

Thank you!

Barbara

galgeek commented:
Hi @whimboo!

I've updated.

Thank you again for all your help!

whimboo commented:
This looks good to me. I rebased the commits and merged the branch as https://github.com/mozilla/testdaybot/commit/e38ec3ae5b32984fe4bf26e90637cdf6cb40f077.
